lib/Dialect/RTL/Ops.cpp

```diff
@@ -6,13 +6,215 @@
+                             resultAttrs))
+    return failure();
+
+  // Record the argument and result types as an attribute.  This is necessary
+  // for external modules.
+  auto type = builder.getFunctionType(argTypes, resultTypes);
+  result.addAttribute(getTypeAttrName(), TypeAttr::get(type));
+
+  // If function attributes are present, parse them.
+  if (parser.parseOptionalAttrDictWithKeyword(result.attributes))
+    return failure();
+
+  assert(argAttrs.size() == argTypes.size());
+  assert(resultAttrs.size() == resultTypes.size());
+
+  auto *context = result.getContext();
+
+  // Postprocess each of the arguments.  If there was no 'rtl.name'
+  // attribute, and if the argument name was non-numeric, then add the
+  // rtl.name attribute with the textual name from the IR.  The name in the
+  // text file is a load-bearing part of the IR, but we don't want the
+  // verbosity in dumps of including it explicitly in the attribute
+  // dictionary.
+  for (size_t i = 0, e = argAttrs.size(); i != e; ++i) {
+    auto &attrs = argAttrs[i];
+
+    // If an explicit name attribute was present, don't add the implicit one.
+    bool hasNameAttr = false;
+    for (auto &elt : attrs)
+      if (elt.first.str() == "rtl.name")
+        hasNameAttr = true;
+    if (hasNameAttr || entryArgs.empty())
+      continue;
+
+    auto &arg = entryArgs[i];
+
+    // The name of an argument is of the form "%42" or "%id", and since
+    // parsing succeeded, we know it always has one character.
+    assert(arg.name.size() > 1 && arg.name[0] == '%' && "Unknown MLIR name");
+    if (isdigit(arg.name[1]))
+      continue;
+
+    auto nameAttr = StringAttr::get(arg.name.drop_front(), context);
+    attrs.push_back({Identifier::get("rtl.name", context), nameAttr});
+  }
+
+  // Add the attributes to the function arguments.
+  addArgAndResultAttrs(builder, result, argAttrs, resultAttrs);
+
+  // Parse the optional function body.
+  auto *body = result.addRegion();
+  if (parser.parseOptionalRegion(
+          *body, entryArgs, entryArgs.empty() ? ArrayRef<Type>() : argTypes))
+    return failure();
+
+  RTLModuleOp::ensureTerminator(*body, parser.getBuilder(), result.location);
+  return success();
+}
+
+FunctionType getRTLModuleOpType(Operation *op) {
+  auto typeAttr =
+      op->getAttrOfType<TypeAttr>(rtl::RTLModuleOp::getTypeAttrName());
+  return typeAttr.getValue().cast<FunctionType>();
+}
+
+static void printRTLModuleOp(OpAsmPrinter &p, Operation *op) {
+  using namespace mlir::impl;
+
+  FunctionType fnType = getRTLModuleOpType(op);
+  auto argTypes = fnType.getInputs();
+  auto resultTypes = fnType.getResults();
+
+  // Print the operation and the function name.
+  auto funcName =
+      op->getAttrOfType<StringAttr>(::mlir::SymbolTable::getSymbolAttrName())
+          .getValue();
+  p << op->getName() << ' ';
+  p.printSymbolName(funcName);
+
+  printFunctionSignature(p, op, argTypes, /*isVariadic*/ false, resultTypes);
```

Comment:
nit: `/*isVariadic*/` -> `/*isVariadic=*/`; and maybe add comments for the `resultTypes` as well?